text/0000-variance-and-implicit-interfaces.md

```diff
@@ -0,0 +1,129 @@
+
+With the new syntax, this could have been defined more simply as:
+
+```pony
+type ElementsAny is Elements[+Data]
+```
+
+This simple type alias would be equivalent to the interface:
+
+```pony
+interface val ElementsAny
+  fun string(): String
+  fun size(): USize
+  fun apply(i: USize): Data?
+  fun values(): Iterator[Data]
+```
+
+If the use site had instead used the contravariant syntax (`Elements[-Data]`), a different set of methods would be included in the interface:
+
+```pony
+interface val ElementsAny
+  fun string(): String
+  fun size(): USize
+  fun ref push(elem: Data)
+```
+
+Or, if the type argument had been omitted (`Elements[_]`), only the methods not using the `A` type parameter at all would remain:
+
+```pony
+interface val ElementsAny
+  fun string(): String
+  fun size(): USize
+```
+
+You might guess that the type argument omission syntax may not be very useful in the real world, because it prevents you from using any methods where the type parameter is in the method signature. However, in the seven examples of this `XAny` pattern that I studied from the body of my own libraries, six of them would be satisfied using the type argument omission interface instead of covariant or contravariant interfaces. These are often cases where the type parameter is used to parameterize some implementation detail of the generic type, but the generic type still implements a set of useful methods where the type parameter doesn't come into play as part of the signature (that is, the signatures are identical in the general case).
+
+# How We Teach This
+
+The tutorial chapter on generics would need a new page to explain these concepts and give some practical examples of use.
+
+# How We Test This
+
+Add compiler sugar tests that verify the anonymous interface that was invoked into existence at the use site gets the proper set of methods in every case.
+
+# Drawbacks
+
+* Adds a new concept and new syntax to represent it in the language. However, I think the burden of learning new syntax is lessened by the following:
+    * The `+`/`-` syntax to represent covariance and contravariance is fairly similar to its use in other languages, like Scala
+    * The `_` syntax for omission should be quite intuitive, given that it's used in pattern matching and some other places as the "don't care" symbol. In this position it means "Give me the interface for when I don't care what the type argument was".
+
+* Adds some new complexity into the compiler to implement and maintain in the type system.
+
+# Alternatives
+
+- Create the anonymous interface as a trait instead, so that other types not mentioned at the use site won't end up as "accidentally" implementing the anonymous interace.
```

Comment:
Nit: this wouldn't actually be a trait, but some mix between the two (an interface which requires the class name to be the same).

Reply:
I was thinking of it as an anonymous trait that was implicitly created, as well as being implicitly applied to your class, so that your class was the only type that was nominally linked to the trait.

I think this way of thinking about it is easier to understand than saying the the compiler checks the class name, but maybe others feel differently.